OpenRA.Game/Translation.cs

```diff
@@ -0,0 +1,121 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Fluent.Net;
+using Fluent.Net.RuntimeAst;
+using OpenRA.FileSystem;
+
+namespace OpenRA
+{
+	public class Translation
+	{
+		readonly IEnumerable<MessageContext> messageContexts;
+
+		public Translation(string language, string[] translations, IReadOnlyFileSystem fileSystem)
+		{
+			if (translations == null || !translations.Any())
+				return;
+
+			messageContexts = GetMessageContext(language, translations, fileSystem);
+		}
+
+		IEnumerable<MessageContext> GetMessageContext(string language, string[] translations, IReadOnlyFileSystem fileSystem)
+		{
+			var backfall = translations.Where(t => t.EndsWith("en.ftl"));
+			var paths = translations.Where(t => t.EndsWith(language + ".ftl"));
+			foreach (var path in paths.Concat(backfall))
+			{
+				var stream = fileSystem.Open(path);
+				using (var reader = new StreamReader(stream))
+				{
+					var options = new MessageContextOptions { UseIsolating = false };
+					var messageContext = new MessageContext(language, options);
+					var errors = messageContext.AddMessages(reader);
+					foreach (var error in errors)
+						Log.Write("debug", error.ToString());
+
+					yield return messageContext;
+				}
+			}
+		}
+
+		public string GetFormattedMessage(string key, IDictionary<string, object> args = null, string attribute = null)
+		{
+			if (key == null)
+				return "";
+
+			foreach (var messageContext in messageContexts)
```

Comment:
Beware that `GetMessageContext` will be called each time we use `messageContexts`